## Question :

Berapa rata-rata dan minimal delivery fee restoran bagi tiap customer

## Answer :

Berdasarkan data set food delivery maka dapat dicari rata-rata dan minimal delivery fee yang harus dikeluarkan customer

## Algoritma :

1. Menjawab tujuan pertanyaan tersebut melalui suatu bentuk program
2. Mengimport library yang akan digunakan 
3. Mengimport data set food delivery
4. Menentukan variabel yang digunakan, yaitu cust_id, delivery_fee
5. Menentukan mean dan minimal delivery_fee per cust_id dengan aggregation

## 1.Purpose

Pertanyaan tersebut mengacu pada ongkos kirim yang dibebankan pada customer. Sehingga akan dicari rata-rata dan juga ongkos kirim minimal berdasarkan data set food delivery yang telah tersedia

## 2.Import Library

In [1]:
import pandas as pd
import numpy as np

## 3.Get Data

In [2]:
PATH_DATA = '../mydata/'

In [3]:
df_csv = pd.read_csv(PATH_DATA+'food_delivery_datasets.csv')

In [4]:
df_csv

order_id  voucher_amount                   date_time  eta_seconds  \
0     FOOD-10455               0  2018-08-12T20:35:08.001006         1800   
1     FOOD-10985               0  2018-09-23T17:44:12.243273         2173   
2     FOOD-10911               0  2018-09-19T18:08:09.387991         1418   
3     FOOD-10932               0  2018-09-21T01:24:35.272499         1903   
4     FOOD-10961               0  2018-09-22T17:57:45.523161         1716   
...          ...             ...                         ...          ...   
1071  FOOD-10921               0  2018-09-20T15:01:52.152828          900   
1072  FOOD-10265            5000  2018-07-26T14:43:23.655294         1579   
1073  FOOD-10534               0  2018-08-21T20:10:59.283185         2606   
1074  FOOD-10158               0  2018-07-16T14:42:23.214427         2428   
1075  FOOD-10770               0  2018-09-11T14:18:19.494258         1547   

      food_price  cust_id  rating  delivery_fee   resto_id     status  \
0          18900  1089034     5.0          4000  742638033  Completed   
1         149000   180146     5.0         11000  742767245  Completed   
2          43800   366528     NaN          9000  742618303  Completed   
3          39900   511038     5.0         14000  742541172  Completed   
4          77700  1220358     NaN          4000  742604875  Completed   
...          ...      ...     ...           ...        ...        ...   
1071      103200  1296335     5.0          4000  742742150  Completed   
1072      145700  1135466     5.0          7000  742566851  Completed   
1073      177200  1383625     NaN         15000  742571888  Completed   
1074       49300  1094787     5.0          9000  742725064  Completed   
1075       43000  1010922     5.0          9000  742594499  Completed   

      random_id  
0      0.511149  
1      0.832936  
2      0.828026  
3      0.239377  
4      0.995587  
...         ...  
1071   0.248328  
1072   0.268251  
1073   0.132528  
1074   0.167946  
1075   0.358051  

[1076 rows x 11 columns]

In [5]:
## mengambil sampel data sebanyak 20 transaksi
df_csv.sample(20)

order_id  voucher_amount                   date_time  eta_seconds  \
1027  FOOD-10026               0  2018-07-03T18:11:16.851186         1622   
550   FOOD-10091               0  2018-07-10T10:25:03.806353         3078   
195   FOOD-10121               0  2018-07-12T12:30:31.465527          846   
486   FOOD-10503               0  2018-08-18T15:13:45.414424         3100   
1008  FOOD-10863               0  2018-09-16T18:04:29.548916         1790   
929   FOOD-10352               0  2018-08-03T20:22:08.328983          920   
530   FOOD-10223            5000  2018-07-22T13:10:45.480457         2341   
68    FOOD-10764               0  2018-09-10T20:59:45.659483         3198   
994   FOOD-10424               0  2018-08-10T19:37:13.988449         1918   
817   FOOD-10136               0  2018-07-14T16:03:13.336066         1384   
280   FOOD-10006            3000  2018-07-01T17:19:33.935851         1269   
980   FOOD-10829           15000  2018-09-15T13:21:28.741573         1056   
476   FOOD-10240            6000  2018-07-23T17:55:41.060742         1627   
544   FOOD-10059            3000  2018-07-06T19:33:16.782827           86   
1036  FOOD-10763               0  2018-09-10T20:46:40.632116         1436   
676   FOOD-10350               0  2018-08-03T19:58:58.983436         1590   
55    FOOD-10109               0  2018-07-11T12:49:00.138757           60   
448   FOOD-10721               0  2018-09-07T16:27:51.678802          461   
429   FOOD-10746               0  2018-09-09T14:57:35.449857         1903   
276   FOOD-10461               0  2018-08-13T15:06:56.896798         2151   

      food_price  cust_id  rating  delivery_fee   resto_id     status  \
1027      107500  1351793     5.0         13000  742539606  Completed   
550        35100  1347550     5.0          8000  742675191  Completed   
195        33300   854144     5.0          2000  742528179  Completed   
486        56300  1094787     5.0          4000  742758916  Completed   
1008       62400  1319474     5.0         10000  742531018  Completed   
929        30500   884440     5.0          4000  742549466  Completed   
530        63000  1089034     5.0          3000  200005639  Completed   
68        155100  1379865     NaN         10000  742735098  Completed   
994        73100   616763     5.0          5000  742559783  Completed   
817        72100   410472     NaN         25000  742591683  Cancelled   
280        33300   316834     5.0         10000  742554025  Completed   
980        39200    82690     5.0             0  742710437  Completed   
476        40400   616763     5.0             0  742595557  Completed   
544       214700  1472725     NaN         13000  742596786  Cancelled   
1036       39800   357568     5.0          9000  742627192  Completed   
676        28700   943758     5.0          9000  742632178  Completed   
55         27500  1284060     NaN          2000  742753560  Cancelled   
448        48300   366528     NaN         18000  200009033  Cancelled   
429        71200  1479204     NaN          4000  742531591  Completed   
276       147800  1360716     5.0         11000  742743527  Completed   

      random_id  
1027   0.800681  
550    0.957514  
195    0.884955  
486    0.815216  
1008   0.886628  
929    0.539374  
530    0.228238  
68     0.136865  
994    0.401823  
817    0.869667  
280    0.216188  
980    0.978142  
476    0.478615  
544    0.285282  
1036   0.298974  
676    0.316932  
55     0.376058  
448    0.969484  
429    0.518762  
276    0.833032

In [6]:
df_csv.max()

order_id                          FOOD-11076
voucher_amount                         25000
date_time         2018-09-30T20:17:40.796955
eta_seconds                             6082
food_price                            448500
cust_id                              1548571
rating                                     5
delivery_fee                           44000
resto_id                           742814735
status                             Completed
random_id                           0.999124
dtype: object

Berdasarkan output df_csv.max() diatas, maka diketahui bahwa transaksi dengan delivery_fee 44000 (paling mahal) berstatus 'Completed'.

In [7]:
df_csv.min()

order_id                          FOOD-10001
voucher_amount                             0
date_time         2018-07-01T11:36:10.581264
eta_seconds                                2
food_price                             12900
cust_id                                22895
rating                                     1
delivery_fee                               0
resto_id                           200000639
status                             Cancelled
random_id                          0.0016161
dtype: object

Berdasarkan output df_csv.min() diatas, maka diketahui bahwa tarnsaksi dengan delivery_fee 0 (free ongkir) berstatus 'Cancelled'.

## 4.Determine Variable

In [8]:
## variabel yang digunakan adalah cust_id dan delivery_fee
df_csv[['cust_id','delivery_fee']]

cust_id  delivery_fee
0     1089034          4000
1      180146         11000
2      366528          9000
3      511038         14000
4     1220358          4000
...       ...           ...
1071  1296335          4000
1072  1135466          7000
1073  1383625         15000
1074  1094787          9000
1075  1010922          9000

[1076 rows x 2 columns]

Data diatas merupakan data dengan kolom yang menyatakan cust_id dan juga delivery_fee yang diperoleh berdasarkan data set food delivery.

#### Delivery Fee for Each Order

In [9]:
## delivery_fee per order sudah terdapat di data
df_csv[['delivery_fee']]

delivery_fee
0             4000
1            11000
2             9000
3            14000
4             4000
...            ...
1071          4000
1072          7000
1073         15000
1074          9000
1075          9000

[1076 rows x 1 columns]

In [10]:
## mendeskripsikan delivery_fee untuk memperoleh berbagai nilai yang akan dicari
df_csv[['delivery_fee']].describe()

delivery_fee
count   1076.000000
mean    9252.788104
std     6723.595497
min        0.000000
25%     4000.000000
50%     9000.000000
75%    12000.000000
max    44000.000000

In [11]:
## dapat juga dicari satu per satu
## Count
df_csv['delivery_fee'].count()

1076

In [12]:
## Mean
df_csv['delivery_fee'].mean()

9252.788104089219

In [13]:
## Standar Deviasi
df_csv['delivery_fee'].std()

6723.595496852201

In [14]:
## Min
df_csv['delivery_fee'].min()

0

In [15]:
## Max
df_csv['delivery_fee'].max()

44000

Nilai-nilai diatas merupakan nilai yang menerangkan delivery_fee tiap orderan (bukan per customer). Berdasarkan nilai tersebut maka diketahui bahwa rata-rata delivery_fee untuk per orderan adalah 9252.788104089219 dan minimal delivery_fee untuk per orderan adalah 0 (gratis ongkos kirim).

## 5.Aggregation 

#### Delivery Fee for Each Customer (each cust_id)

In [16]:
## Akan dicari rata-rata delivery_fee dan minimal delivery_fee restoran yang dibebankan pada customer, maka perlu dilakukan aggregation diantara cust_id dan delivery_fee yaitu dengan groupby.

In [17]:
## memyatukan atau grouping customer Id
data_group_cust = df_csv.groupby('cust_id')

In [18]:
data_group_cust['cust_id'].size()

cust_id
22895       2
59565       3
62551       3
70701       3
82690       5
           ..
1510407     1
1514562    13
1542763     4
1544551     5
1548571     6
Name: cust_id, Length: 138, dtype: int64

In [19]:
## memyatukan atau grouping cust_id dengan delivery_fee
data_group_id = data_group_cust['delivery_fee'].mean().reset_index()
data_group_id

cust_id  delivery_fee
0      22895   9000.000000
1      59565  11333.333333
2      62551   8333.333333
3      70701   9000.000000
4      82690   2000.000000
..       ...           ...
133  1510407  16000.000000
134  1514562  10538.461538
135  1542763   5500.000000
136  1544551  11600.000000
137  1548571   8833.333333

[138 rows x 2 columns]

In [20]:
## mendeskripsikan delivery_fee berdasarkan hasil group_id
data_group_id[['delivery_fee']].describe()

delivery_fee
count    138.000000
mean    8879.748957
std     4964.288862
min        0.000000
25%     5446.428571
50%     8916.666667
75%    10937.500000
max    30000.000000

In [21]:
## dapat juga dicari satu per satu
## Count untuk data_group_id
data_group_id['delivery_fee'].count()

138

In [22]:
## Mean untuk data_group_id
data_group_id['delivery_fee'].mean()

8879.748956751362

In [23]:
## Standar Deviasi untuk data_group_id
data_group_id['delivery_fee'].std()

4964.288861644278

In [24]:
## Min untuk data_group_id
data_group_id['delivery_fee'].min()

0.0

In [25]:
## Max untuk data_group_id
data_group_id['delivery_fee'].max()

30000.0

Nilai-nilai diatas merupakan nilai yang menerangkan delivery_fee tiap per customer. Berdasarkan nilai tersebut maka diketahui bahwa rata-rata delivery_fee yang dibebankan pada tiap customer adalah 8879.748956751362 dan minimal delivery_fee untuk tiap customer adalah 0 (gratis ongkos kirim).